TST: Apply ufunc signature and type test fixmes.

Kai-Striega opened the pull request:
Update tests in ``test_ufuncs`` where some tests were labelled
``fixme``. In particular this commit addresses two case:

1. Invalid ufunc signature tests now test for value errors directly
rather than via a try..except clause and have more descriptive test
names.

2. Tests for generic loops now test for all types to be tested.

Kai-Striega committed:
TST: Apply ufunc signature and type test fixmes

Update tests in ``test_ufuncs`` where some tests were labelled
``fixme``. In particular this commit addresses two case:

1. Invalid ufunc signature tests now test for value errors directly
rather than via a try..except clause and have more descriptive test
names.

2. Tests for generic loops now test for all types to be tested.

seberg commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -111,67 +111,76 @@ def test_generic_loops(self):
         # check unary PyUFunc_f_f.
         msg = "PyUFunc_f_f"
-        x = np.zeros(10, dtype=np.single)[0::2]
+        x = np.zeros(10, dtype=np.single)
```
Have not checked carefully, but I am confused by the removal of the `[::2]`. This probably has a purpose here to ensure a non-contiguous array, I think; which is important to test that SIMD optimization doesn't take a wrong path.

Kai-Striega replied:
Thanks for bringing up ``[0::2]`` part.

I'm not sure if what ``[0::2]`` is supposed to test. The non-contiguous array is plausible, but even if it is, the test still past without it - so whatever ``[::2]`` tests is *probably* not being caught by the tests. Even if it is, would you expect SIMD to change the *dtypes* of the input or output arrays?

I'll add ``[::2]`` back in iff you think it's the better idea, but then I would like to find an instance where a test actually fails without it.

seberg replied:
This does check equality of arrays/values on top of equality of data types,  no? I see a lot of assert almost equal.

Kai-Striega replied:
It tests ``abs(desired-actual) < 1.5 * 10**(-decimal)``. Looking at the (docs)[https://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.assert_almost_equal.html] there is actually a warning to avoid using this in favour of ``assert_allclose``. Perhaps I should change these?

Kai-Striega replied:
And no ``assert_almost_equal`` does **not** check the arrays datatype

mattip commented:
It should be possible to rewrite `test_generic_loops` more compactly using `pytest.mark.parametrize`

Kai-Striega commented:
@mattip I'll be happy to rewrite it. I've been looking at #8300, but haven't contributed to the main NumPy repo yet (except some docs) - so I wanted to start small before rewriting any big changes.

Kai-Striega commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -44,6 +44,103 @@ def test_extobj_refcount(self):
+        def logical_xor(self, obj):
+            return np.bool_(1)
+
+    def test_unaryPyUFunc_O_O(self):
```
todo: ``unaryPyUFunc`` -> ``unary_PyUFunc``

Kai-Striega commented:
@mattip I've parameterized the tests and moved them into their own class. 

@seberg I have not yet fixed the test for contiguous arrays, but plan on adding them back in to keep the status quo.

I've also replaced the ``assert_almost_equal`` with ``assert_allclose`` (as recommended in the [docs])(https://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.assert_almost_equal.html) and explicitly testing for the data type. Perhaps this should be its own commit.

Kai-Striega committed:
TST: Parametrize ufunc generic loop tests.

Tests for ufunc generic loops follow the same structure for unary and binary
functions, excpet for the datatypes to be tested. This commit
parametrizes the tests, removing code shared by each of the unary and
binary function tests respectively.

Kai-Striega committed:
TST: Tighten test of generic Ufunc loops.

The previous assertion tested was ``assert_almost_equal`` which has two
major drawbacks:

* ``assert_almost_equal`` does not handle floating point values
consistently, however many of the test compare floating point results.
As recommended in the docs test has changed to use ``assert_allclose``
which provides better floating point comparisons.

* It fails to check if the expected dtypes are returned. Currently the
tests check how *close* the resultant numbers are to one another, not,
if the dtypes are the same. The updated tests explicitly compare dtypes

Kai-Striega commented:
@mattip I've added the ``[::2]`` checks back in as requested by @seberg and split the previous commit into two - one for parametrizing and one for tightening the tests - which is why I've had to force push.

All the tests are passing and the original *fixmes* are complete. I'm still looking at #8300, I think any further changes should be in another PR. Is there anything else before we may merge?

mattip commented:
Thanks @Kai-Striega
